File: Interviews/GFG/random-problems/Medium/Maximum-of-all-subarrays-of-size-k/Maximum-of-all-subarrays-of-size-k.cpp
class Solution {
public:
  vector <int> max_of_subarrays(int *arr, int n, int k) {
    vector<int> ans(n - k + 1);
    deque<int> dq;
    for (int i = 0; i < n; i++) {
      while (!dq.empty() and arr[i] > arr[dq.back()]) dq.pop_back();
      dq.push_back(i);
      while (!dq.empty() and dq.front() < (i - k + 1)) dq.pop_front();
      if (i >= k - 1) ans[i - k + 1] = arr[dq.front()];
    }

    return ans;
  }
};

File: Interviews/GFG/Daily-Problems/2022/October/Day-25/Maximum-of-all-subarrays-of-size-k.py
from collections import deque
class Solution:
    def max_of_subarrays (self, arr, n, k):
        dq, ans = deque(),[]
        for i in range(n):
            while len(dq) and arr[i] > arr[dq[-1]]: dq.pop()
            dq.append(i)
            while len(dq) and dq[0] < (i - k + 1): dq.popleft()
            if i >= k - 1: ans.append(arr[dq[0]]);
        
        return ans